src/viewport.cpp

```diff
@@ -200,8 +227,9 @@ void DeleteWindowViewport(Window *w)
 {
 	if (w->viewport == nullptr) return;
 
+	container_unordered_remove(_viewport_window_cache, w->viewport);
```

Comment:
`_viewport_window_cache` seems to be an ordered container to me. Any reason for this extra function instead of the normal `std::erase/std::remove` idiom?

Reply:
Also, I think automatic memory management with `std::unique_ptr` is preferred where possible.

Reply:
I didn't write this code; please tell me how you would fix it?

Reply:
I will figure out what those two things mean and then try again. Thanks for the feedback.

Reply:
Removing an element from a vector at an unknown location is even on wikipedia: https://en.wikipedia.org/wiki/Erase–remove_idiom

Right now these viewport structs are managed with manual new/delete, which is error-prone as a spot where a delete is needed can always be missed. `std::unique_ptr` will automatically manage the memory. Using it would entail figuring out who the actual owner should be though, which can sometimes make it hard to really use well.

Reply:
_viewport_window_cache is not an ordered container. The order in which it is iterated is not important and so the swap and pop idiom is fine. It wouldn't make much difference either way really.

The existing GUI code is almost entirely implemented using manual new/delete or malloc/free. Updating struct Window/InitializeWindowViewport/DeleteWindowViewport wouldn't be difficult though.

Reply:
>  `std::unique_ptr` will automatically manage the memory. Using it would entail figuring out who the actual owner should be though.

Without looking into the code, I would assume the new `_viewport_window_cache` would be the logical owner of the viewports. That way, it can possibly also automatically manage their allocation and destruction (if I correctly understand how std::vector works).
I was about to dive into the code to figure out if this assumption holds true.

Reply:
`struct Window` is the owner. `_viewport_window_cache` is, as per the name, just a cache.

> Without looking into the code
You really should look at the code. The diff won't make sense without the context of the code it is going into.

Reply:
@JGRennison `std::vector` is not an unordered container, at least in the STL sense. It's a sequential contains and thus works fine with `erase/remove` and doesn't need any custom algorithm. A standard remove algorithm for true unordered or associative containers was only introduced with C++20 though.

Reply:
@JGRennison yep, I can see that now. I was trying to be too clever for myself, and got a nice SIGSEGV for my troubles :wink:

Reply:
@michicc perhaps https://github.com/OpenTTD/OpenTTD/pull/8217/commits/e27dd6a7d00448df34eeca12b500ee216c2791e3 is more to your tastes?

(For the sake of testing and learning the intricacies of smart pointers, I simplified `DeleteWindowViewport()` to its bare minimum. In actual implementation, I intend to mirror the erase/remove idiom after the viewport's smart pointer is deleted with `w->viewport = nullptr;`)

Reply:
Is there any particular reason why the viewport cache shouldn't be managed by the constructor/destructor of the `ViewportData` (or even `ViewPort`) `struct`?

Sidenote: is there any good reason why one is spelled "ViewPort" and the other is spelled "Viewport"?